### PR TITLE
LPS-200511 Implement saving of hierarchical object fields in DSM

### PR DESCRIPTION
**Pull request by markocikos**
### References

- [LPS-195914 Allow users to save the mapped hierarchical object fields](https://issues.liferay.com/browse/LPS-195914)
- Requires: https://github.com/liferay-frontend/liferay-portal/pull/3770

### What is the goal of this PR?

In this PR, we are implementing saving of hierarchical object fields. The UI and handling is significantly more complex, as UI is based on `ClayTreeView`, with a matching tree data structure. This means recursive traversal for each action that modifies the tree. 

See technical notes in comments.

### What does it look like?

To showcase new behavior, I created two Objects, and a relationship between them:
- `Person`
- `City`
- `livesIn` relationship, 1:N `City`: `Person`. Multiple people live in a city, obviously :slightly_smiling_face: 

Note that the hierarchical fields in the end are empty, and that's expected. FDS doesn't yet support new format of names, and `livesIn` field is an objects, defaulting to an empty field.

https://github.com/liferay-frontend/liferay-portal/assets/5435511/41df899c-f010-45dc-b865-58511bcc6acb

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts [hunk 1]
```diff
@@ -78,7 +81,7 @@ function getValidFields({
 		fields.push({
 			format: properties[propertyKey].format,
 			label: propertyKey,
-			name: propertyKey,
+			name: `${contextPath}${propertyKey}`,
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/api.ts:84]: `contextPath` here is just a local internal mechanism, to build hierarchical `name`, delimited by a dot.

I considered extending `FDSField` persistence to include `contextPath`, but that approach quickly increased in complexity. We are using `name` as a unique identifier in a lot of places. All those places would need to be changed to check both `name` and `contextPath`. On the other hand, complex names maintain their uniqueness, and I have not found any blocker to using them.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx [hunk 2]
```diff
@@ -8,48 +8,160 @@ import {TreeView} from '@clayui/core';
...
+
+function visit(fields: Array<IFieldTreeItem>, callback: Function) {
+	fields.forEach((field) => {
+		callback(field);
+
+		if (field.children) {
+			visit(field.children, callback);
+		}
+	});
+}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx:35]: I stole this from [AssetCategoryTree](https://github.com/liferay/liferay-portal/blob/master/modules/apps/asset/asset-categories-item-selector-web/src/main/resources/META-INF/resources/js/AssetCategoryTree.es.js#L29-L37). Shoutout to @victorg1991 :sparkles:

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx [hunk 3]
```diff
@@ -8,48 +8,160 @@ import {TreeView} from '@clayui/core';
...
+			selectedKeys.add(savedFDSField.name);
+
+			field.savedId = savedFDSField.id;
+		}
+
+		field.id = field.name;
+	});
+
+	return [selectedKeys, fields];
+};
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx:62]: There was a problem here with the `id` property. It is used by `ClayTreeView` as a unique identifier, in its internal selection mechanism. In our case, ID is set only if the field is saved to DSM persistence. If it's a new field taken from OpenAPI, it does not have an ID. That's why it cannot be used as a unique identifier. Instead, we are setting `ID` property to have `name` value, as `name` is unique. This is local change, caused by and specific only to the `ClayTreeView` component.

Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx:62]: It sounds strange having a couple of similar lists with different types (`deletionIds`) :shrug: [here](https://github.com/liferay-frontend/liferay-portal/pull/3806/files#diff-3bdfde06fddcbc2f5c626b00d67c7727b00ae4df1d5ca558a02f71edd78aaf50R162) and [here](https://github.com/liferay-frontend/liferay-portal/pull/3806/files#diff-3bdfde06fddcbc2f5c626b00d67c7727b00ae4df1d5ca558a02f71edd78aaf50R162) but as you mention, the scope is limited.

**Review by juanjofgliferay — APPROVED** [review]
LGTM :rocket:

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
The failure in https://github.com/liferay-frontend/liferay-portal/pull/3806#issuecomment-1808115075 is relevant, a bug. Working on a fix

**Comment by markocikos** [comment]
ci:forward

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx (hunk 3 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/fds_view/fields/modal_content/AddFieldsModalContent.tsx:62]: @juanjofgliferay Your intuition was right on this one! The inconsistency resulted in the bug from https://github.com/liferay-frontend/liferay-portal/pull/3806#issuecomment-1808246312, where we ended up comparing strings and integers. My bad for not looking into this more thoroughly. Also, a lesson for me: don't be lazy with setting `Function` as TS type, argument and return types would have prevented this.

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
ci:stop

**Comment by markocikos** [comment]
Force-pushed with a small fix, regression from https://github.com/liferay-frontend/liferay-portal/pull/3770. [Relevant Slack discussion](https://liferay.slack.com/archives/C3JBR21HA/p1699957626114059).

**Comment by markocikos** [comment]
ci:forward

**Comment by markocikos** [comment]
ci:forward:force

**Pull request closed without merging**